### PR TITLE
upgrade jest

### PR DESCRIPTION
**Pull request by brad-decker**
## Explanation
Fixes issue related to slack reports of unit tests consuming upwards of more than 20gb of memory and 4gb of swap. Note that this PR does nothing to fix the underlying memory leaks in our tests and that should be explored separately. This works around the memory issue by forcing worker processes to terminate once they've consumed 500mb of memory. I chose 500mb because our ci environment has 4gb of ram and we run two workers when running tests in jest. This would allow us to theoretically add another 6 workers without running into out of memory issues. I'll test that theory in another PR aimed at improving CI. 

Consequences of this PR:
1. The default environment for tests is node, so had to install the jsdom environment and specify jsdom as our testEnvironment in jest.config.js. This can be changed *per file*. 
2. Had to temporarily disable the button.stories.test.js file because it eventually imports a .mdx file and the transform that was previously working for 'mdx' files in jest no longer works. 
3. Had to patch the types for jsdom to not use the `import { type TypeName}` syntax
4. Removed the yargs patch because the newest version checks for the existence of Error.createErrorBoundary before calling it.
5. Some minor syntax changes

## Pre-Merge Checklist

- [x] PR template is filled out
- [ ] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added
- [ ] PR is linked to the appropriate GitHub issue
- [ ] PR has been added to the appropriate release Milestone

### + If there are functional changes:

- [ ] Manual testing complete & passed
- [ ] "Extension QA Board" label has been applied

### CODE REVIEW

**Review by tmashuang — DISMISSED** [review]
LGTM, will re-approve once the merge conflicts are resolved.

**Review by tmashuang — DISMISSED** [review]

File: ui/components/ui/button/button.stories.skippedtest.js
Comment (georgewrmarshall) [inline ui/components/ui/button/button.stories.skippedtest.js:7]: I think we can probably remove this file altogether. It was added in an attempt to check if a story broke because of a component update. I think storybook snapshots would be a more scalable solution with less overhead.

**Review by georgewrmarshall — COMMENTED** [review]
As per our slack conversation I think we can probably remove the `ui/components/ui/button/button.stories.skippedtest.js` file altogether

**Review by georgewrmarshall — APPROVED** [review]
LGTM!

File: jest.config.js
Comment (mcmire) [inline jest.config.js:59]: `multiformats`, `uuid`, and `nanoid` makes sense, because they have recently(ish) converted to ESM, and I've seen Jest choke on these packages in other projects.

We don't ship ESM code in `@metamask/controllers` or `@metamask/snap-controllers`, however, so why did we have to list these here?

File: patches/@types+jsdom++parse5+7.0.0.patch
Comment (mcmire) [inline patches/@types+jsdom++parse5+7.0.0.patch:7]: Sorry why do we have to remove `type` here?

File: jest.config.js
Comment (brad-decker) [reply jest.config.js:59]: some of the dependencies nested *inside* the metamask/controllers repo are ESM and i could not figure out how to specify that recursion in this ignore pattern.

Comment (mcmire) [reply jest.config.js:59]: Ah I see. Interesting...

File: patches/@types+jsdom++parse5+7.0.0.patch
Comment (mcmire) [reply patches/@types+jsdom++parse5+7.0.0.patch:7]: Ohh... wow. That's subtle. How do other people even use this package then 🤔

**Comment by mcmire** [comment]
Had to do a bit of research on this one. Looks like the next version of Jest includes `workerIdleMemoryLimit`, and that version isn't out yet, so this PR bumps it to 29.alpha. The [docs for this option](https://jestjs.io/docs/next/configuration#workeridlememorylimit-numberstring) link to [this issue](https://github.com/facebook/jest/issues/11956) which explains where it came from. It also might be worth following [this PR](https://github.com/facebook/jest/pull/12205) to see if that fixes things further. Or [this issue in Node](https://github.com/nodejs/node/issues/44211).

**Review by mcmire — APPROVED** [review]
Makes sense to me.

File: patches/@types+jsdom++parse5+7.0.0.patch
Comment (brad-decker) [reply patches/@types+jsdom++parse5+7.0.0.patch:7]: no idea. unless it has something to do with our setup.

**Comment by brad-decker** [comment]
@mcmire my apologies for not posting more details in the description. You ended up doing some double work that I could have possibly saved some time for you if I would have simply transcribed @Gudahtt and I's thread in the extension channel into the PR description. I'll do better in the future.

**Pull request merged**
